ignite/handlers/checkpoint.py

```diff
@@ -257,6 +257,11 @@ def _check_lt_n_saved(self, or_equal=False):
             return True
         return len(self._saved) < self._n_saved + int(or_equal)
 
+    def _will_be_full(self):
```

Comment:
Can't we reuse `_check_lt_n_saved` instead of `_will_be_full` ?

Reply:
No, I don't think so because we need to explicitly check equality. If the next element will overflow the queue (i.e., `if len(self._saved) == self._n_saved` before the update), then we need to pop (if necessary). Since `_check_lt_n_saved` only checks for lt or lte. This code will hopefully eventually be replaced by some `heapq` implementation, so inconsistency (in private API) is unfortunate but not the worst.

Reply:
Well, not exactly. We can append new item (line 319), then check if need to drop the worst checkpoint and finally save new checkpoint

Reply:
Isn't that exactly what this PR seeks to fix? 😄

Reply:
This PR seeks to fix the order of operations without introducing redundant functions :)

Reply:
The process you just described (append, check, drop) is exactly what is in master now and has been deemed problematic because it admits a case of `n+1` elements in `Checkpoint._saved`. I might be misunderstanding what you mean, but checking first if the `_saved` is going to overflow requires an equality check (not <=).

Reply:
Let me give a simplified example of the code from above link and what I think we would like

<details>

```python
from collections import namedtuple


_n_saved = 3
_saved = []


def _check_lt_n_saved(or_equal=False):
    if _n_saved is None:
        return True
    return len(_saved) < _n_saved + int(or_equal)


Item = namedtuple("Item", ["priority", "filename"])

for priority in [0, 1, 2, 1, 3, 5, 6, 7]:
    print("\n-", priority)

    if _check_lt_n_saved() or _saved[0].priority < priority:

        filename = "check_{}".format(priority)
        if any(item.filename == filename for item in _saved):
            print("Found the same filename")
            continue

        # When we setup the checkpoint, Checkpoint's state dict should be up to date as we add new item
        # and remove optionally other items.
        _saved.append(Item(priority, filename))        
        _saved.sort(key=lambda item: item[0])
        print("-- Appended", priority, ". Now _saved: ", [i.priority for i in _saved])

        if not _check_lt_n_saved(or_equal=True):
            item = _saved.pop(0)
            print("-- Removed ", item.priority, ". Now _saved", [i.priority for i in _saved])
#             if isinstance(self.save_handler, BaseSaveHandler):            
#                 self.save_handler.remove(item.filename)

#         checkpoint = self._setup_checkpoint()
#         self.save_handler(checkpoint, filename)
        print("-- Saved ", filename, ". Now _saved", [i.priority for i in _saved])
```

</details>

Output

<details>

```
- 0
-- Appended 0 . Now _saved:  [0]
-- Saved  check_0 . Now _saved [0]

- 1
-- Appended 1 . Now _saved:  [0, 1]
-- Saved  check_1 . Now _saved [0, 1]

- 2
-- Appended 2 . Now _saved:  [0, 1, 2]
-- Saved  check_2 . Now _saved [0, 1, 2]

- 1
Found the same filename

- 3
-- Appended 3 . Now _saved:  [0, 1, 2, 3]
-- Removed  0 . Now _saved [1, 2, 3]
-- Saved  check_3 . Now _saved [1, 2, 3]

- 5
-- Appended 5 . Now _saved:  [1, 2, 3, 5]
-- Removed  1 . Now _saved [2, 3, 5]
-- Saved  check_5 . Now _saved [2, 3, 5]

- 6
-- Appended 6 . Now _saved:  [2, 3, 5, 6]
-- Removed  2 . Now _saved [3, 5, 6]
-- Saved  check_6 . Now _saved [3, 5, 6]

- 7
-- Appended 7 . Now _saved:  [3, 5, 6, 7]
-- Removed  3 . Now _saved [5, 6, 7]
-- Saved  check_7 . Now _saved [5, 6, 7]
```

</details>

Make senses ?

Reply:
No, it doesn't make sense because 

> Now _saved:  [0, 1, 2, 3]

means you have `len(_saved) > n_saved` even if it's temporary. This is exactly what we have now.

Reply:
> Even if this 'checkpoint' isn't saved to disk, it still puts the Checkpoint object in an inconsistent state.

Inconsistent in sense of temporary `len(_saved) > n_saved` , right ?

Reply:
If it is the only remark, you can change the order of "append" and "remove":
```python
        # When we setup the checkpoint, Checkpoint's state dict should be up to date as we add new item
        # and remove optionally other items.
        if not _check_lt_n_saved(or_equal=False):
            item = _saved.pop(0)
            print("-- Removed ", item.priority, ". Now _saved", [i.priority for i in _saved])
#             if isinstance(self.save_handler, BaseSaveHandler):            
#                 self.save_handler.remove(item.filename)
    
        _saved.append(Item(priority, filename))        
        _saved.sort(key=lambda item: item[0])
        print("-- Appended", priority, ". Now _saved: ", [i.priority for i in _saved])
```
and it will give : 
```
- 3
-- Removed  0 . Now _saved [1, 2]
-- Appended 3 . Now _saved:  [1, 2, 3]
-- Saved  check_3 . Now _saved [1, 2, 3]

- 5
-- Removed  1 . Now _saved [2, 3]
-- Appended 5 . Now _saved:  [2, 3, 5]
-- Saved  check_5 . Now _saved [2, 3, 5]

- 6
-- Removed  2 . Now _saved [3, 5]
-- Appended 6 . Now _saved:  [3, 5, 6]
-- Saved  check_6 . Now _saved [3, 5, 6]

- 7
-- Removed  3 . Now _saved [5, 6]
-- Appended 7 . Now _saved:  [5, 6, 7]
-- Saved  check_7 . Now _saved [5, 6, 7]
```